lib/codegen/meta/src/cdsl/instructions.rs

```diff
@@ -0,0 +1,112 @@
+use super::types::ValueType;
+
+pub struct InstructionList {
+    groups: Vec<InstructionGroup>,
+}
+
+impl InstructionList {
+    pub fn new() -> Self {
+        Self { groups: Vec::new() }
+    }
+
+    pub fn group(&mut self, name: &'static str, f: impl FnOnce(&mut InstructionGroup)) {
+        let mut group = InstructionGroup::new(name);
+        f(&mut group);
+        self.groups.push(group);
+    }
+}
+
+pub struct InstructionGroup {
+    name: &'static str,
+    insts: Vec<Instruction>,
+}
+
+impl InstructionGroup {
+    fn new(name: &'static str) -> Self {
+        Self {
+            name,
+            insts: Vec::new(),
+        }
+    }
+
+    pub fn add(&mut self, inst: Instruction) {
+        self.insts.push(inst);
+    }
+}
+
+pub struct Instruction {
+    name: &'static str,
+    doc: &'static str,
+    ins: &'static [()],
+    outs: &'static [()],
+    constraints: &'static [Constraint],
```

Comment:
We can make these owned `String`s and `Vec`s, because this is meta code so we don't need to micro-manage allocations, and so that meta code can dynamically build things like operand lists when needed.